### ISSUE TITLE
Implement common interface in applyconfigs to make them identifiable as such

### ISSUE THREAD

**Issue by alvaroaleman** [kubernetes/kubernetes#118138]
### What would you like to be added?

Make all generated applyconfigs implement some common interface that makes them identifiable as such, similiar to `runtime.Object`. How exactly that interface looks doesn't matter, it could be a simple `IsApplyConfig() bool`.

### Why is this needed?

In order to be able to offer a good UX for SSA, dynamic clients such as controller-runtimes need a first-class `Apply` or similiar method. In order for that method to be easy to use, it needs to only accept applyconfigs. Because applyconfigs do not implement a common interface, this isn't possible today and such a method needs to accept `any`.

**Comment by apelisse** [comment]
cc @Jefftree 

Agree with the general idea, also +1 on John's idea. We're definitely thinking about this.

**Comment by Jefftree** [comment]
Now that we're adopting the same generator as client-go, I agree it makes sense to share the interface between the client-go applyconfiguration types and custom generated types. At least in k/k, we never had a common interface for applyconfigurations because each type has their own typed `Apply()` function so this is something new we'll have to define.

As for the interface, there are a couple routes, but a simple `IsApplyConfig() bool` is probably not enough as we may need to access the getters and setters for things like TypeMeta and ObjectMeta. controller-runtime most likely needs some of those functions and I'd probably recommend an interface that enumerates the functions that are need. Eg: (note, this not an exhaustive list)

```
type ApplyConfiguration interface {
  WithName()
  WithNamespace()
  WithKind()
  WithAPIVersion()
  GetName()
  GetNamespace()
  GetKind()
  GetAPIVersion()
}
```

Also +1 on the idea of generics, though not really sure how the implementation would look in practice.

**Comment by Jefftree** [comment]
> Good point regarding also needing basic object meta and apiVersion and Kind information.
> 
> Would it make sense to hardcode the apiVersion and Kind information inside the generated type? They are generated for a specific api in a specific version, so being able to set that to potentially something else doesn't seem very useful

Yes, that's already part of the generator. I see it's also included in the [PR](https://github.com/kubernetes-sigs/controller-tools/blob/4affe2086a475bff012ebd76ab93e7578f5ed132/pkg/applyconfigurations/testdata/cronjob/applyconfiguration-master/testdata/cronjob/cronjob.go#L22-L29). If the values are already set, maybe we don't need the setters, but some of the getters are still needed to determine which endpoint to send a request to.

**Comment by erikgb** [comment]
I am not a Go expert, but I would really love to see some movement on this. SSA has been around for a while now, and still no first-class support in controller-runtime for it. 😢 Isn't adding setters from applyconfigurations to the new interface blocked by https://github.com/golang/go/issues/49085? And with just the getters, which should be fine, we'll probably end up with a super-interface of `meta.Object`.

**Comment by erikgb** [comment]
I would love to see some movement on this issue! Would it make sense if I created a draft PR to start more detailed discussions about this new common interface for all apply configs?

Looking at migration to SSA in multiple operators I'm involved in, I think it would be nice if this new interface is implemented by `unstructured.Unstructured`. Why? A common pattern we use when an operator manages resources in an external (3. party) API, is to represent the desired state (fields we care for) of the owned/controlled resource using Go templates. So we sacrifice type-safety to avoid pulling in external (3. party) operator dependencies - potentially conflicting with our K8s dependencies.

**Comment by erikgb** [comment]
Adding a breadcrumb to https://github.com/kubernetes/kubernetes/issues/113773 - seems related to me.

**Comment by erikgb** [comment]
I am dreaming of getting some movement on this issue, so I created https://github.com/kubernetes/kubernetes/pull/124659 hoping to start some detailed discussions. PTAL!

**Comment by caozhuozi** [comment]
Im also dreaming of getting some movement on this issue like @erikgb 😢😢😢
